crates/core/src/live_socket/navigation/mod.rs

```diff
@@ -0,0 +1,259 @@
+    /// Returns the current history ID if changed
+    pub fn navigate(&mut self, url: Url, opts: NavOptions, emit_event: bool) -> Option<HistoryId> {
+        let action = opts.action.clone();
+        let next_dest = self.speculative_next_dest(&url, opts.state.clone());
+        let next_id = next_dest.id;
+        let event = NavEvent::new_from_navigate(next_dest.clone(), self.current(), opts);
+
+        match self.handle_event(event, emit_event) {
+            HandlerResponse::Default => {}
+            HandlerResponse::PreventDefault => return None,
+        };
+
+        match action {
+            Some(NavAction::Replace) => self.replace_entry(next_dest),
+            None | Some(NavAction::Push) => self.push_entry(next_dest),
+        }
+
+        // successful navigation invalidates previously coalesced state from
+        // calls to `back`
+        self.future.clear();
+        Some(next_id)
+    }
+
+    // Returns true if the navigator can go back one entry.
+    pub fn can_go_back(&self) -> bool {
+        self.history.len() >= 2
+    }
+
+    // Returns true if the navigator can go forward one entry.
+    pub fn can_go_forward(&self) -> bool {
+        !self.future.is_empty()
+    }
+
+    // Returns true if the `id` is tracked in the navigation context.
+    pub fn can_traverse_to(&self, id: HistoryId) -> bool {
+        let hist = self.history.iter().find(|ent| ent.id == id);
+        let fut = self.future.iter().find(|ent| ent.id == id);
+        hist.or(fut).is_some()
+    }
+
+    // Returns all of the tracked history entries by cloning them.
+    // They are in traversal sequence order, with no guarantees about
+    // the position of the current entry.
+    pub fn entries(&self) -> Vec<NavHistoryEntry> {
+        self.history
+            .iter()
+            .chain(self.future.iter().rev())
+            .cloned()
+            .collect()
+    }
+
+    /// Calls the handler for reload events
+    pub fn reload(&mut self, info: Option<Vec<u8>>, emit_event: bool) -> Option<HistoryId> {
+        let current = self.current()?;
+        let id = current.id;
+        let event = NavEvent::new_from_reload(current, info);
+
+        match self.handle_event(event, emit_event) {
+            HandlerResponse::Default => {}
+            HandlerResponse::PreventDefault => return None,
+        };
+
+        Some(id)
+    }
+
+    /// Navigates back one step in the stack, returning the id of the new
+    /// current entry if successful.
+    /// This function fails if there is no current
+    /// page or if there are no items in history and returns [None].
+    pub fn back(&mut self, info: Option<Vec<u8>>, emit_event: bool) -> Option<HistoryId> {
+        if !self.can_go_back() {
+            log::warn!("Attempted `back` navigation without at minimum two entries.");
+            return None;
+        }
+
+        let previous = self.current()?;
+
+        let next = self.history[self.history.len() - 2].clone();
```

Comment:
Hmm. I feel like this should be `self.history.get(self.history.len() - 2)?.clone()`.

Reply:
the bounds check is validated by `!self.can_go_back()`  on 117

Reply:
I figured you'd validate the bounds but wasn't sure where it was.